### PR TITLE
EOS-26012: Performance Stats: Write a listener to read from perfstat topic when REST Call will come from consumer

### PR DESCRIPTION
**Pull request by Pranali04796**
# Problem Statement
- EOS-26012: Performance Stats: Write a listener to read from perfstat topic when REST Call will come from consumer [EOS:26102](https://jts.seagate.com/browse/EOS-26762) [EOS-26946](https://jts.seagate.com/browse/EOS-26946)

# Design
-  Expose GET API (https://{url}:{port}/**api/v2/metrics/stats/perf**) to consume metrics present in message bus.
- Currently we are reading from a static file /opt/seagate/cortx/csm/templates/metrics.txt
- After reading sending those metrics to message bus and consuming them to expose
- We can manually push metrics to message bus and consume using GET api provided

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [x] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [x] Testing was performed with RPM
CSM-setup will create related topic:
 image
**Test1: consume GET api.**
 By deafult api will read metric from a static file:  **/opt/seagate/cortx/csm/templates/metrics.txt** after reading it will push them to message bus later consume them from message bus to expose them.
 image

**Test2: Produce metrics and consume them using GET api**
Produce messages using script below and consume them using GET api 
(It will expose both metrics produced by script and which are present in static file ):

**from cortx.utils.message_bus import MessageBus, MessageProducer 
from cortx.utils.conf_store import Conf
Conf.load('cortx_conf', 'yaml:///etc/cortx/cluster.conf')
message_server_endpoints = Conf.get('cortx_conf', 'cortx>external>kafka>endpoints')
MessageBus.init(message_server_endpoints)
message_list = ["This is test message 1",  "This is test message 2"]
producer = MessageProducer(producer_id='sspl_sensor', message_type='perf_stat', method='sync')
producer.send(message_list)** 
image

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

-----

### CODE REVIEW

File: csm/plugins/cortx/convertor.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
+from abc import ABCMeta, abstractmethod
```
Comment (pawankumarsrivastava) [inline csm/plugins/cortx/convertor.py:1]: remove this. Not used anywhere

File: csm/core/services/stats.py [hunk 2]
```diff
@@ -160,6 +176,37 @@ async def get_metrics(self, stats_id, metrics_list, from_t, to_t, interval,
...
+        for metric in StatsAppService.BUFFER:
+            # Add new line while parsing if metrics are not read from a static file
+            metrics += str(metric)
+        return web.Response(text=metrics)
+
+    async def get_perf_metrics(self):
+        """ get performace stat metrics"""
+        StatsAppService.BUFFER = []
+        # Read from a static file and publish to msassage bus
+        with open("/opt/seagate/cortx/csm/templates/metrics.txt", "r") as fp:
```
Comment (pawankumarsrivastava) [inline csm/core/services/stats.py:200]: Is this file checked in? Have you made sure this file is available in the same location?

File: csm/core/services/stats.py [hunk 3]
```diff
@@ -160,6 +176,37 @@ async def get_metrics(self, stats_id, metrics_list, from_t, to_t, interval,
...
+            while True:
+                line = fp.readline()
+                if not line:
+                    break
+                self.metrics_client.send([line])
```
Comment (pawankumarsrivastava) [inline csm/core/services/stats.py:205]: Please have a different method for sending stats and call it here

**Review by pawankumarsrivastava — CHANGES_REQUESTED** [review]

File: csm/core/services/stats.py (hunk 2 above)
Comment (Pranali04796) [reply csm/core/services/stats.py:200]: yes. this file is already present.

File: csm/plugins/cortx/convertor.py (hunk 1 above)
Comment (Pranali04796) [reply csm/plugins/cortx/convertor.py]: removed

File: csm/core/services/stats.py (hunk 3 above)
Comment (Pranali04796) [reply csm/core/services/stats.py]: Added new method for sending metric.

**Review by pawankumarsrivastava — APPROVED** [review]

**Pull request merged**